Implemented PXB-3097 - Remove Conditional version parsing from PXB code

altmannmarcelo opened the pull request:
https://jira.percona.com/browse/PXB-3097

PXB 8.0 had to implement various condition-based code blocks to deal with new versions of redo or bugs in previous versions.

The below commits were removed:

64fbda51
fef769df
4ce04308
aef77a61 - Since #1023 we correctly write `last_checkpoint_lsn` via `log_status` in `stop_at`.
34d83f7c
841cd3b1
d972829f

storage/innobase/include/log0constants.h
Removed older LOG_HEADER_CREATOR
Removed pxb-2455.sh as it is no longer relevant.

altmannmarcelo committed:
Implemented PXB-3097 - Remove Conditional version parsing from PXB code

https://jira.percona.com/browse/PXB-3097

PXB 8.0 had to implement various condition-based code blocks to deal
with new versions of redo or bugs in previous versions.

The below commits were removed:

64fbda51
fef769df
4ce04308
aef77a61 - Since #1023 we correctly write `last_checkpoint_lsn` via
`log_status` in `stop_at`.
34d83f7c
841cd3b1
d972829f

storage/innobase/include/log0constants.h
Removed older LOG_HEADER_CREATOR
Removed pxb-2455.sh as it is no longer relevant.

Added version check at --prepare. As part of this work, we now also
save the server flavor as part of xtrabackup_info file.

Removed dd_upgrade_sdi test as it is not relevant anymore. Since
upgrading from previous versions is not supported anymore.
The test is redundant with duplicate_sdi and can be removed.

satya-bodapati commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -57,57 +54,34 @@ Redo_Log_Reader::Redo_Log_Reader() {
 }
 
 bool Redo_Log_Reader::find_start_checkpoint_lsn() {
-  if (log_sys->m_files.ctx().m_files_ruleset == Log_files_ruleset::CURRENT) {
```
Can you add an assert here so that a new enum type (Log_files_ruleset) added will cause a crash or compilation failure?

altmannmarcelo replied:
Done

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -4938,13 +4931,13 @@ static bool xtrabackup_init_temp_log(void) {
   }
 
-  if (log_format < Log_format::VERSION_8_0_1) {
+  if (log_format < Log_format::CURRENT) {
```
Upstream didn't bump the redo log header version right? :thinking: So the redo log version is same across 8.0.34 and 8.1?

altmannmarcelo replied:
correct, I will adjust as this potentially allows for 8.1 to prepare 8.0 backup

altmannmarcelo replied:
Done

satya-bodapati reviewed (commented):
LGTM! Just few questions

satya-bodapati commented:
Jenkins results please

satya-bodapati commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -1812,13 +1814,14 @@ char *get_xtrabackup_info(MYSQL *connection) {
-                     incremental_lsn,                /* innodb_from_lsn */
-                     metadata_to_lsn,                /* innodb_to_lsn */
-                     (xtrabackup_tables              /* partial */
+                     mysql_server_version_comment.c_str(), /* server_flavor */
```
are we writing directly into std::string internal buffer directly? I think, this is not a recommended way.

altmannmarcelo replied:
no, this part is printing out the content of mysql_server_version_comment as char*

satya-bodapati commented on storage/innobase/xtrabackup/src/xtrabackup.cc:
```diff
@@ -2502,9 +2503,19 @@ static bool xtrabackup_read_info(char *filename) {
       xtrabackup::utils::get_version_number(mysql_server_version_str);
 
+  /* prepare version check */
+  if (!check_server_version(xb_server_version, mysql_server_version_str,
```
I think it is better to put this in prepare path. this is more like read or parse function can be reused at other places.

satya-bodapati commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -1886,6 +1889,7 @@ bool write_xtrabackup_info(MYSQL *connection) {
                  "tool_version VARCHAR(255) DEFAULT NULL,"
                  "ibbackup_version VARCHAR(255) DEFAULT NULL,"
                  "server_version VARCHAR(255) DEFAULT NULL,"
+                 "server_flavor VARCHAR(255) DEFAULT NULL,"
```
what if the table exists already (came from 8.0 server that this table), do we add column via ALTER?

altmannmarcelo replied:
Do we need to care about this on major upgrades?

altmannmarcelo replied:
I think we should. I do have a back log ticket to improve this check, will just add the alter

satya-bodapati commented on storage/innobase/xtrabackup/src/backup_mysql.cc:
```diff
@@ -1792,6 +1793,7 @@ char *get_xtrabackup_info(MYSQL *connection) {
                      "tool_version = %s\n"
                      "ibbackup_version = %s\n"
                      "server_version = %s\n"
+                     "server_flavor = %s\n"
```
This is strictly not required for version check? Any reason to add this to the history table?

altmannmarcelo replied:
Same info is inserted into history table and xtrabackup_info file (same function)

altmannmarcelo committed:
Adjusted bootstrap

Added retries to bootstrap to handle network issues from jenkins.
Added python3 version of subunit2junitxml

altmannmarcelo commented:
After a lot of Jenkins fixes - https://pxb.cd.percona.com/job/percona-xtrabackup-8.1-test-param/15/

satya-bodapati commented:
@altmannmarcelo I see on pxb crash from the test results. Can you please check and fix this?

```
2023-09-08T13:32:16.440537-00:00 0 [Note] [MY-011825] [Xtrabackup] Starting to parse redo log at lsn = 19033159
2023-09-08T13:32:16Z UTC - mysqld got signal 6 ;
Most likely, you have hit a bug, but this error can also be caused by malfunctioning hardware.
BuildID[sha1]=
Thread pointer: 0x0
Attempting backtrace. You can use the following information to find out
where mysqld died. If you see no messages after this, something went
terribly wrong...
stack_bottom = 0 thread_stack 0x100000
/tmp/results/PXB/bin/xtrabackup(my_print_stacktrace(unsigned char const*, unsigned long)+0x41) [0x190ef21]
/tmp/results/PXB/bin/xtrabackup(print_fatal_signal(int)+0x3bc) [0xdf141c]
/tmp/results/PXB/bin/xtrabackup(handle_fatal_signal+0x95) [0xdf14c5]
/lib64/libc.so.6(+0x54df0) [0x7f44bc410df0]
/lib64/libc.so.6(+0xa158c) [0x7f44bc45d58c]
/lib64/libc.so.6(raise+0x16) [0x7f44bc410d46]
/lib64/libc.so.6(abort+0xd3) [0x7f44bc3e47f3]
/tmp/results/PXB/bin/xtrabackup() [0x89caa6]
/tmp/results/PXB/bin/xtrabackup() [0x8a4f87]
/tmp/results/PXB/bin/xtrabackup(Redo_Log_Writer::write_buffer(unsigned char*, unsigned long)+0x1a4) [0x900cf4]
/tmp/results/PXB/bin/xtrabackup(Redo_Log_Data_Manager::start()+0xcd) [0x9028ed]
/tmp/results/PXB/bin/xtrabackup(xtrabackup_backup_func()+0x94b) [0x8ca49b]
/tmp/results/PXB/bin/xtrabackup(main+0x1a0e) [0x86c9be]
/lib64/libc.so.6(+0x3feb0) [0x7f44bc3fbeb0]
/lib64/libc.so.6(__libc_start_main+0x80) [0x7f44bc3fbf60]
/tmp/results/PXB/bin/xtrabackup(_start+0x25) [0x8991d5]

Please report a bug at https://jira.percona.com/projects/PXB
2023-09-08 13:32:16: run.sh: ===> xtrabackup failed with exit code 2
```

satya-bodapati commented:
https://pxb.cd.percona.com/job/percona-xtrabackup-8.1-test-param/15/CMAKE_BUILD_TYPE=RelWithDebInfo,DOCKER_OS=oraclelinux%3A9,XTRABACKUP_TARGET=innodb80/testReport/junit/(root)/suites_compression_pxb-1901/sh/

satya-bodapati commented:
@altmannmarcelo code wise, I don't have any other suggestions

altmannmarcelo commented:
This comes from trunk - happens on RHEL9 and we need to investigate - https://jira.percona.com/browse/PXB-3149

satya-bodapati commented:
@altmannmarcelo Thanks for creating the bug (that exists on 8.1 trunk). Ideally, we should squash the commits but it seems there are other Jenkins-related commits in this PR. Maybe squash them into two commits? (or create different PR?). How do you want to handle this?

altmannmarcelo commented:
@satya-bodapati Same PR :D

altmannmarcelo commented:
@satya-bodapati grouped them into two.
